src/transformStateWithClones.js

```diff
@@ -7,7 +7,31 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let currentState = { ...state };
+  const statesHistory = [];
+
+  actions.forEach((action) => {
+    switch (action.type) {
+      case 'clear':
+        currentState = {};
+        break;
+      case 'addProperties':
+        currentState = { ...currentState, ...action.extraData };
+        break;
+      case 'removeProperties':
+        currentState = { ...currentState };
```

Comment:
[DONT REPEAT YOURSELF] - The creation of a new copy of `currentState` inside the 'removeProperties' case is unnecessary because you already have a fresh copy from the beginning of the forEach loop. You can remove this line to avoid redundant code.